File: frontend/src/lib/targeting.ts
/**
 * Unified targeting — the client side of the shared audience/search model.
 *
 * `Targeting` is "the kind of person we're after": edited in the campaign composer's audience card
 * AND in Find People (same <TargetingEditor>), stored on `Campaign.criteria`, posted to
 * `/people/search`. `evaluateFit` is a byte-for-byte mirror of the backend `app/targeting.py`
 * `evaluate()`, so the composer's live "~N match" estimate agrees with what server-side ranking
 * produces. Keep the two in lockstep — backend tests/test_targeting.py pins the canonical cases.
 *
 * Two axes: `evaluateFit` → Fit 0-100 (weighted fraction of *specified* criteria matched; skills
 * fractional; excludes hard-disqualify; "in the audience" at >= FIT_THRESHOLD) and `reachability`
 * → whether we can act on them. Fit is NOT affected by having an email. Titles/companies match on
 * substring ("VP" ⊇ "SVP"); excludes on word boundaries ("intern" ≠ "International"); seniority/
 * function via a synonym taxonomy. technologies/keywords stay search-only but give a neutral floor
 * when they're the only criteria.
 */
export const FIT_THRESHOLD = 40;

export interface Targeting {
  titles: string[];
  seniorities: string[];
  functions: string[];
  skills: string[];
  locations: string[];
  companies: string[];
  industries: string[];
  company_sizes: string[];
  technologies: string[];
  keywords: string;
  exclude_companies: string[];
  exclude_titles: string[];
}

export function emptyTargeting(): Targeting {
  return {
    titles: [],
    seniorities: [],
    functions: [],
    skills: [],
    locations: [],
    companies: [],
    industries: [],
    company_sizes: [],
    technologies: [],
    keywords: "",
    exclude_companies: [],
    exclude_titles: [],
  };
}

/** Coerce a stored `Campaign.criteria` (partial / legacy {titles,skills,locations}) into Targeting. */
export function toTargeting(c: Partial<Targeting> | null | undefined): Targeting {
  return { ...emptyTargeting(), ...(c ?? {}), keywords: c?.keywords ?? "" };
}

export type ChipField = Exclude<keyof Targeting, "keywords">;

/** True if the targeting specifies anything at all (any chip field or free-text keywords). */
export function targetingHasCriteria(t: Targeting): boolean {
  if (t.keywords.trim().length > 0) return true;
  return (Object.keys(t) as (keyof Targeting)[]).some(
    (k) => k !== "keywords" && (t[k] as string[]).length > 0,
  );
}

export const TARGETING_FIELDS: {
  key: ChipField;
  label: string;
  group: "person" | "company" | "exclude";
  scored: boolean;
  placeholder: string;
}[] = [
  { key: "titles", label: "Titles", group: "person", scored: true, placeholder: "VP of Sales…" },
  { key: "seniorities", label: "Seniority", group: "person", scored: false, placeholder: "Director, VP…" },
  { key: "functions", label: "Function", group: "person", scored: false, placeholder: "Engineering, Sales…" },
  { key: "skills", label: "Skills", group: "person", scored: true, placeholder: "Salesforce…" },
  { key: "locations", label: "Locations", group: "person", scored: true, placeholder: "EU, Berlin…" },
  { key: "companies", label: "Companies", group: "company", scored: true, placeholder: "Acme, Globex…" },
  { key: "industries", label: "Industries", group: "company", scored: true, placeholder: "Fintech…" },
  { key: "company_sizes", label: "Company size", group: "company", scored: true, placeholder: "51-200, 501-1,000…" },
  { key: "technologies", label: "Technologies", group: "company", scored: false, placeholder: "React, AWS…" },
  { key: "exclude_companies", label: "Exclude companies", group: "exclude", scored: true, placeholder: "Competitor Inc…" },
  { key: "exclude_titles", label: "Exclude titles", group: "exclude", scored: true, placeholder: "Intern, Student…" },
];

// ---- scoring (mirror of app/targeting.py — keep byte-for-byte) ----

const WEIGHTS = {
  titles: 30,
  skills: 30,
  companies: 20,
  seniorities: 20,
  industries: 15,
  locations: 15,
  functions: 10,
  company_sizes: 10,
};

const SEARCH_ONLY_FLOOR = 50;

const REGION_ALIASES: Record<string, string[]> = {
  eu: ["de", "uk", "nl", "pt", "ie", "fr", "es", "it", "remote · eu"],
  us: ["us", "usa", "united states"],
  remote: ["remote"],
};

const SENIORITY_ALIASES: Record<string, string> = {
  "vice president": "vp", vp: "vp", svp: "vp", evp: "vp",
  "c-level": "exec", "c-suite": "exec", cxo: "exec", chief: "exec",
  ceo: "exec", cto: "exec", cfo: "exec", coo: "exec",
  founder: "exec", owner: "exec", president: "exec", partner: "exec",
  director: "director", dir: "director", head: "director",
  senior: "senior", sr: "senior",
  lead: "lead", staff: "lead", principal: "lead",
  manager: "manager", mgr: "manager",
  mid: "mid", intermediate: "mid",
  junior: "junior", jr: "junior", entry: "junior", intern: "junior", associate: "junior",
};
const FUNCTION_ALIASES: Record<string, string> = {
  engineering: "engineering", eng: "engineering", software: "engineering",
  developer: "engineering", development: "engineering", dev: "engineering",
  sales: "sales",
  marketing: "marketing", growth: "marketing",
  product: "product", design: "design", data: "data",
  operations: "operations", ops: "operations",
  finance: "finance", accounting: "finance",
  people: "people", hr: "people", "human resources": "people", recruiting: "people",
  support: "support", "customer success": "support",
  legal: "legal",
};

export type Reachability = "verified" | "reachable" | "needs_enrichment";

export interface FitContact {
  title?: string | null;
  skills?: string[];
  location?: string | null;
  email?: string | null;
  email_status?: string | null;
  linkedin_url?: string | null;
  company?: string | null;
  industry?: string | null;
  company_size?: string | null;
  seniority?: string | null;
  function?: string | null;
}
export interface FitResult {
  score: number;
  matched: boolean;
  reasons: string[];
}

/** Permissive substring (so "VP" matches "SVP"). Used for positive title/company matches. */
function containsAny(value: string | null | undefined, needles: string[]): boolean {
  const v = (value ?? "").toLowerCase();
  return !!v && needles.some((n) => n && v.includes(n.toLowerCase()));
}

const isAlnum = (ch: string): boolean => /[\p{L}\p{N}]/u.test(ch);

/** Precise whole-word containment (so "intern" ≠ "International"). Used for excludes. */
function wordContains(value: string | null | undefined, needle: string): boolean {
  const v = (value ?? "").toLowerCase();
  const n = needle.toLowerCase().trim();
  if (!v || !n) return false;
  let start = 0;
  for (;;) {
    const i = v.indexOf(n, start);
    if (i < 0) return false;
    const before = i > 0 ? v[i - 1] : "";
    const after = i + n.length < v.length ? v[i + n.length] : "";
    if (!isAlnum(before) && !isAlnum(after)) return true;
    start = i + 1;
  }
}

function anyWord(value: string | null | undefined, needles: string[]): boolean {
  return needles.some((n) => n && wordContains(value, n));
}

function canon(value: string | null | undefined, aliases: Record<string, string>): string {
  const k = (value ?? "").toLowerCase().trim();
  return aliases[k] ?? k;
}

/** Normalized-synonym equality for the single-token fields (seniority / function). */
function bucketMatch(
  value: string | null | undefined,
  crits: string[],
  aliases: Record<string, string>,
): boolean {
  if (crits.length === 0) return false;
  const cv = canon(value, aliases);
  if (!cv) return false;
  const set = new Set(crits.map((c) => canon(c, aliases)));
  return set.has(cv);
}

function locationMatches(loc: string | null | undefined, crits: string[]): boolean {
  if (crits.length === 0) return true;
  const cl = (loc ?? "").toLowerCase();
  return crits.some((c) => {
    const k = c.toLowerCase();
    if (cl.includes(k)) return true;
    return (REGION_ALIASES[k] ?? []).some((tok) => cl.includes(tok));
  });
}

export function evaluateFit(c: FitContact, t: Targeting): FitResult {
  if (anyWord(c.company, t.exclude_companies) || anyWord(c.title, t.exclude_titles)) {
    return { score: 0, matched: false, reasons: ["excluded by targeting"] };
  }

  const want = (t.skills ?? []).map((s) => s.toLowerCase());
  const have = (c.skills ?? []).map((s) => s.toLowerCase());
  const overlap = want.filter((w) => have.some((h) => wordContains(h, w)));

  const titleMatch = containsAny(c.title, t.titles);
  const companyMatch = containsAny(c.company, t.companies);
  const industryMatch = containsAny(c.industry, t.industries);
  const sizeMatch = containsAny(c.company_size, t.company_sizes);
  const locMatch = locationMatches(c.location, t.locations);
  const senMatch = bucketMatch(c.seniority, t.seniorities, SENIORITY_ALIASES);
  const fnMatch = bucketMatch(c.function, t.functions, FUNCTION_ALIASES);

  const cats: { weight: number; hit: number }[] = [];
  if (t.titles.length) cats.push({ weight: WEIGHTS.titles, hit: titleMatch ? 1 : 0 });
  if (want.length) cats.push({ weight: WEIGHTS.skills, hit: overlap.length / want.length });
  if (t.companies.length) cats.push({ weight: WEIGHTS.companies, hit: companyMatch ? 1 : 0 });
  if (t.seniorities.length) cats.push({ weight: WEIGHTS.seniorities, hit: senMatch ? 1 : 0 });
  if (t.industries.length) cats.push({ weight: WEIGHTS.industries, hit: industryMatch ? 1 : 0 });
  if (t.locations.length) cats.push({ weight: WEIGHTS.locations, hit: locMatch ? 1 : 0 });
  if (t.functions.length) cats.push({ weight: WEIGHTS.functions, hit: fnMatch ? 1 : 0 });
  if (t.company_sizes.length) cats.push({ weight: WEIGHTS.company_sizes, hit: sizeMatch ? 1 : 0 });

  const totalW = cats.reduce((s, x) => s + x.weight, 0);
  let fit: number;
  if (totalW > 0) fit = (100 * cats.reduce((s, x) => s + x.weight * x.hit, 0)) / totalW;
  else if (t.technologies.length || t.keywords.trim()) fit = SEARCH_ONLY_FLOOR;
  else fit = 0;
  const score = Math.min(100, Math.round(fit));

  const reasons: string[] = [];
  if (overlap.length) reasons.push(`matches ${overlap.join(", ")}`);
  if (titleMatch) reasons.push("title fits the role");
  if (senMatch) reasons.push("seniority fits");
  if (fnMatch) reasons.push("right function");
  if (companyMatch) reasons.push("target company");
  if (industryMatch) reasons.push("target industry");
  if (t.locations.length && locMatch) reasons.push("in target location");
  if (sizeMatch) reasons.push("company size fits");
  if (reasons.length === 0) reasons.push("limited overlap with the criteria");

  return { score, matched: score >= FIT_THRESHOLD, reasons };
}

/** A separate axis from fit: can we act on this candidate? Mirror of `reachability()`. */
export function reachability(c: FitContact): Reachability {
  if (c.email_status === "valid") return "verified";
  if (c.email || c.linkedin_url) return "reachable";
  return "needs_enrichment";
}

File: backend/app/targeting.py
"""Unified targeting — ONE audience/search definition, used two ways across the whole app.

A `Targeting` is "the kind of person we're after." It is consumed by exactly two operations:
  * search — each provider adapter maps it onto its own query DSL (Find People / discovery).
  * score  — `evaluate()` measures how well a Contact (or a provider hit) fits it. This drives
             campaign ranking, the audience estimate, single enroll, and Find People result scores.

It is stored on `Campaign.criteria` (the audience) and posted to `/people/search` (the search), so
the two are the same object. The frontend mirrors `evaluate()` byte-for-byte in
`src/lib/targeting.ts` (so the composer's live estimate agrees with what server-side ranking
produces) — keep them in lockstep; `tests/test_targeting.py` pins the canonical cases.

Two orthogonal axes:
  * `evaluate` → Fit 0-100: the weighted fraction of the *specified, scorable* criteria a contact
    matches (each specified field shares a 100-point budget by weight; skills score fractionally).
    An exclude match hard-disqualifies. "In the audience" at `>= FIT_THRESHOLD`. Fit is NOT affected
    by whether we have an email.
  * `reachability` → whether we can act on them (verified email / reachable / needs enrichment).

Titles/companies match on permissive substring (so "VP" ⊇ "SVP"); excludes match on precise
word boundaries (so "intern" ≠ "International"); seniority/function match via a synonym taxonomy.
The remaining search-only fields (`technologies / keywords`) narrow a provider search and, when
they're the *only* criteria, contribute a neutral floor rather than a zero.
"""

from typing import Protocol

from pydantic import BaseModel

from app.core.types import JsonObject

FIT_THRESHOLD = 40


class Targeting(BaseModel):
    """Normalized targeting spec. Each adapter maps this onto its own query language; `evaluate`
    scores a Contact against it. Stored as `Campaign.criteria`; the search request extends it."""

    # --- person ---
    titles: list[str] = []
    seniorities: list[str] = []  # search-only
    functions: list[str] = []  # search-only (department)
    skills: list[str] = []
    locations: list[str] = []
    # --- company ---
    companies: list[str] = []
    industries: list[str] = []
    company_sizes: list[str] = []
    technologies: list[str] = []  # search-only
    # --- free text ---
    keywords: str | None = None  # search-only
    # --- negative targeting ---
    exclude_companies: list[str] = []
    exclude_titles: list[str] = []


class Candidate(Protocol):
    """Anything scorable — a persisted Contact or a live provider hit. Read-only attributes."""

    @property
    def title(self) -> str | None: ...
    @property
    def skills(self) -> list[str] | None: ...
    @property
    def location(self) -> str | None: ...
    @property
    def email(self) -> str | None: ...
    @property
    def email_status(self) -> str | None: ...
    @property
    def linkedin_url(self) -> str | None: ...
    @property
    def company(self) -> str | None: ...
    @property
    def industry(self) -> str | None: ...
    @property
    def company_size(self) -> str | None: ...
    @property
    def seniority(self) -> str | None: ...
    @property
    def function(self) -> str | None: ...


# Relative weight per *scorable* field — only fields the user actually specified share the budget
# (so a one-field audience still scores out of 100). Keep in sync with src/lib/targeting.ts.
WEIGHTS: dict[str, int] = {
    "titles": 30,
    "skills": 30,
    "companies": 20,
    "seniorities": 20,
    "industries": 15,
    "locations": 15,
    "functions": 10,
    "company_sizes": 10,
}

# Fit for an audience defined ONLY by search-only fields (technologies/keywords): the provider
# returned this person for that search, so give a neutral baseline instead of a hard 0.
_SEARCH_ONLY_FLOOR = 50

# Region shorthands so "EU" matches "Berlin, DE", "London, UK", etc.
REGION_ALIASES: dict[str, list[str]] = {
    "eu": ["de", "uk", "nl", "pt", "ie", "fr", "es", "it", "remote · eu"],
    "us": ["us", "usa", "united states"],
    "remote": ["remote"],
}

# Normalize common seniority / function synonyms to one bucket so "VP" == "Vice President" == "SVP".
_SENIORITY_ALIASES: dict[str, str] = {
    "vice president": "vp",
    "vp": "vp",
    "svp": "vp",
    "evp": "vp",
    "c-level": "exec",
    "c-suite": "exec",
    "cxo": "exec",
    "chief": "exec",
    "ceo": "exec",
    "cto": "exec",
    "cfo": "exec",
    "coo": "exec",
    "founder": "exec",
    "owner": "exec",
    "president": "exec",
    "partner": "exec",
    "director": "director",
    "dir": "director",
    "head": "director",
    "senior": "senior",
    "sr": "senior",
    "lead": "lead",
    "staff": "lead",
    "principal": "lead",
    "manager": "manager",
    "mgr": "manager",
    "mid": "mid",
    "intermediate": "mid",
    "junior": "junior",
    "jr": "junior",
    "entry": "junior",
    "intern": "junior",
    "associate": "junior",
}
_FUNCTION_ALIASES: dict[str, str] = {
    "engineering": "engineering",
    "eng": "engineering",
    "software": "engineering",
    "developer": "engineering",
    "development": "engineering",
    "dev": "engineering",
    "sales": "sales",
    "marketing": "marketing",
    "growth": "marketing",
    "product": "product",
    "design": "design",
    "data": "data",
    "operations": "operations",
    "ops": "operations",
    "finance": "finance",
    "accounting": "finance",
    "people": "people",
    "hr": "people",
    "human resources": "people",
    "recruiting": "people",
    "support": "support",
    "customer success": "support",
    "legal": "legal",
}


def as_targeting(x: "Targeting | JsonObject | None") -> Targeting:
    """Coerce a stored criteria dict (or None) into a Targeting. Extra/legacy keys are ignored."""
    return x if isinstance(x, Targeting) else Targeting.model_validate(x or {})


def _contains_any(value: str | None, needles: list[str]) -> bool:
    """Permissive: any needle is a case-insensitive substring of value (so 'VP' matches 'SVP')."""
    v = (value or "").lower()
    return bool(v) and any(n.lower() in v for n in needles if n)


def _word_contains(value: str | None, needle: str) -> bool:
    """Precise: needle occurs in value bounded by non-alphanumerics — so 'intern' matches 'intern
    program' but NOT 'international'. Used for excludes so we never over-disqualify."""
    v = (value or "").lower()
    n = needle.lower().strip()
    if not v or not n:
        return False
    start = 0
    while True:
        i = v.find(n, start)
        if i < 0:
            return False
        before = v[i - 1] if i > 0 else ""
        after = v[i + len(n)] if i + len(n) < len(v) else ""
        if not before.isalnum() and not after.isalnum():
            return True
        start = i + 1


def _any_word(value: str | None, needles: list[str]) -> bool:
    return any(_word_contains(value, n) for n in needles if n)


def _canon(value: str | None, aliases: dict[str, str]) -> str:
    k = (value or "").lower().strip()
    return aliases.get(k, k)


def _bucket_match(value: str | None, crits: list[str], aliases: dict[str, str]) -> bool:
    """Normalized-synonym equality for the single-token fields (seniority / function)."""
    if not crits:
        return False
    cv = _canon(value, aliases)
    return bool(cv) and cv in {_canon(c, aliases) for c in crits}


def _location_matches(location: str | None, crits: list[str]) -> bool:
    if not crits:
        return True  # no location filter → neutral
    cl = (location or "").lower()
    for crit in crits:
        k = crit.lower()
        if k in cl or any(tok in cl for tok in REGION_ALIASES.get(k, [])):
            return True
    return False


def evaluate(contact: Candidate, targeting: "Targeting | JsonObject") -> tuple[int, str]:
    """Fit 0-100 — the weighted fraction of the *specified* criteria this contact matches. Whether
    we can reach them is a separate axis (see `reachability`). Mirrors the TS byte-for-byte."""
    t = as_targeting(targeting)

    # Negative targeting → hard disqualify (word-boundary so 'intern' ≠ 'international').
    if _any_word(contact.company, t.exclude_companies) or _any_word(
        contact.title, t.exclude_titles
    ):
        return 0, "excluded by targeting"

    want = [s.lower() for s in t.skills]
    have = [s.lower() for s in (contact.skills or [])]
    overlap = [w for w in want if any(_word_contains(h, w) for h in have)]

    title_match = _contains_any(contact.title, t.titles)
    company_match = _contains_any(contact.company, t.companies)
    industry_match = _contains_any(contact.industry, t.industries)
    size_match = _contains_any(contact.company_size, t.company_sizes)
    loc_match = _location_matches(contact.location, t.locations)
    sen_match = _bucket_match(contact.seniority, t.seniorities, _SENIORITY_ALIASES)
    fn_match = _bucket_match(contact.function, t.functions, _FUNCTION_ALIASES)

    cats: list[tuple[int, float]] = []  # (weight, hit 0..1) per specified scorable field
    if t.titles:
        cats.append((WEIGHTS["titles"], 1.0 if title_match else 0.0))
    if want:
        cats.append((WEIGHTS["skills"], len(overlap) / len(want)))
    if t.companies:
        cats.append((WEIGHTS["companies"], 1.0 if company_match else 0.0))
    if t.seniorities:
        cats.append((WEIGHTS["seniorities"], 1.0 if sen_match else 0.0))
    if t.industries:
        cats.append((WEIGHTS["industries"], 1.0 if industry_match else 0.0))
    if t.locations:
        cats.append((WEIGHTS["locations"], 1.0 if loc_match else 0.0))
    if t.functions:
        cats.append((WEIGHTS["functions"], 1.0 if fn_match else 0.0))
    if t.company_sizes:
        cats.append((WEIGHTS["company_sizes"], 1.0 if size_match else 0.0))

    total_w = sum(w for w, _ in cats)
    if total_w:
        fit = 100 * sum(w * h for w, h in cats) / total_w
    elif t.technologies or (t.keywords or "").strip():
        fit = float(
            _SEARCH_ONLY_FLOOR
        )  # only search-only criteria → the provider match is the signal
    else:
        fit = 0.0
    score_i = min(100, int(fit + 0.5))  # round-half-up to match JS Math.round

    reasons: list[str] = []
    if overlap:
        reasons.append(f"matches {', '.join(overlap)}")
    if title_match:
        reasons.append("title fits the role")
    if sen_match:
        reasons.append("seniority fits")
    if fn_match:
        reasons.append("right function")
    if company_match:
        reasons.append("target company")
    if industry_match:
        reasons.append("target industry")
    if t.locations and loc_match:
        reasons.append("in target location")
    if size_match:
        reasons.append("company size fits")
    if not reasons:
        reasons.append("limited overlap with the criteria")
    return score_i, "; ".join(reasons)


def reachability(contact: Candidate) -> str:
    """Can we act on this candidate, independent of fit: 'verified' (validated email), 'reachable'
    (email or LinkedIn), or 'needs_enrichment'. A separate axis from `evaluate`. Mirrors the TS."""
    if contact.email_status == "valid":
        return "verified"
    if contact.email or contact.linkedin_url:
        return "reachable"
    return "needs_enrichment"
